webbpsf/wfirst.py

```diff
@@ -276,73 +276,116 @@ def _get_fits_header(self, result, options):
         result[0].header['DETECTOR'] = (self.detector, 'Detector selected')
 
 
-class WFI(WFIRSTInstrument):
-    """
-    WFI represents to the to-be-named wide field imager
-    for the WFIRST mission
-
-    WARNING: This model has not yet been validated against other PSF
-             simulations, and uses several approximations (e.g. for
-             mirror polishing errors, which are taken from HST).
-    """
+class WFIPupilController:
     # "The H158, F184 and W149 filters and the grism are mounted with proximate cold pupil masks"
     # from the final draft of the SDT report, page 92, table 3-2
     UNMASKED_PUPIL_WAVELENGTH_MIN, UNMASKED_PUPIL_WAVELENGTH_MAX = 0.760e-6, 1.454e-6
     MASKED_PUPIL_WAVELENGTH_MIN, MASKED_PUPIL_WAVELENGTH_MAX = 1.380e-6, 2.000e-6
 
-    def __init__(self, set_pupil_mask_on=None):
-        """
-        Initiate WFI
-
-        Parameters
-        -----------
-        set_pupil_mask_on : bool or None
-            Set to True or False to force using or not using the cold pupil mask,
-            or to None for the automatic behavior.
-        """
-        pixelscale = 110e-3  # arcsec/px, WFIRST-AFTA SDT report final version (p. 91)
-        super(WFI, self).__init__("WFI", pixelscale=pixelscale)
+    def __init__(self, wfi):
+        self.wfi = wfi
 
-        self._detector_npixels = 4096
-        self._detectors = _load_wfi_detector_aberrations(os.path.join(self._datapath, 'wim_zernikes_cycle7.csv'))
-        assert len(self._detectors.keys()) > 0
-        self.detector = 'SCA01'
+        self._pupil = None
 
         # Paths to the two possible pupils. The correct one is selected based on requested
         # wavelengths in _validate_config()
-        self._unmasked_pupil_path = os.path.join(self._WebbPSF_basepath,
-                                                 'WFIRST_SRR_WFC_Pupil_Mask_Shortwave_2048.fits')
-        self._masked_pupil_path = os.path.join(self._WebbPSF_basepath,
-                                               'WFIRST_SRR_WFC_Pupil_Mask_Longwave_2048.fits')
+        self._unmasked_pupil_path = None
+        self._masked_pupil_path = None
 
         # Flag to en-/disable automatic selection of the appropriate pupil_mask
         self.auto_pupil = True
 
         self._pupil_mask = "AUTO"
         self.pupil_mask_list = ['AUTO', 'COLD_PUPIL', 'UNMASKED']
 
-        self.pupil = self._unmasked_pupil_path
-        if set_pupil_mask_on is not None:
-            if isinstance(set_pupil_mask_on, bool):
+    @property
+    def pupil(self):
+        return self._pupil
+
+    @pupil.setter
+    def pupil(self, value):
+        self._pupil = value
```

Comment:
We need the parameter because setting the class variable to a value will change the value of the class variable instead of the value of the object that its pointing to. For example: 
```
In [2]: class A:
   ...:     _x = None
   ...:     @property
   ...:     def x(self):
   ...:         return self._x
   ...:     @x.setter
   ...:     def x(self, v):
   ...:         self._x = v
   ...:

In [3]: class B:
   ...:     def __init__(self):
   ...:         self.a = A()
   ...:         self.y = self.a.x
   ...:

In [4]: B()
Out[4]: <__main__.B at 0x10d96e7b8>

In [5]: b = B()

In [6]: b.y

In [7]: b.y = 22

In [8]: b.a.x

In [9]: b.a.x = 44

In [10]: b.y
Out[10]: 22

In [11]:
```